File: Flowscape/camera.py
class Camera:
    ZOOM_MIN = 0.1
    ZOOM_MAX = 5.0

    def __init__(self, world_width, world_height):
        self.x = -world_width / 2   # world coords of top-left screen corner
        self.y = -world_height / 2
        self.zoom = 1.0

    def screen_to_world(self, sx, sy):
        wx = sx / self.zoom + self.x
        wy = sy / self.zoom + self.y
        return wx, wy

    def pan(self, dx_screen, dy_screen):
        self.x -= dx_screen / self.zoom
        self.y -= dy_screen / self.zoom

    def zoom_at(self, sx, sy, factor):
        """Zoom keeping screen point (sx, sy) fixed in world space."""
        wx, wy = self.screen_to_world(sx, sy)
        self.zoom = max(self.ZOOM_MIN, min(self.ZOOM_MAX, self.zoom * factor))
        # Recompute offset so the world point stays under cursor
        self.x = wx - sx / self.zoom
        self.y = wy - sy / self.zoom

File: web/src/camera.ts
import type { MapGeometry, Pt } from './types'

// World <-> screen transform for pan/zoom. World units are feet (y grows
// downward, matching the backend); screen units are canvas CSS pixels.
// This is view-only state — the camera never touches world data.
export class Camera {
  scale = 0.5 // px per foot
  cx = 0 // world x at the canvas center
  cy = 0 // world y at the canvas center

  toScreen(wx: number, wy: number, w: number, h: number): Pt {
    return [(wx - this.cx) * this.scale + w / 2,
            (wy - this.cy) * this.scale + h / 2]
  }

  toWorld(sx: number, sy: number, w: number, h: number): Pt {
    return [(sx - w / 2) / this.scale + this.cx,
            (sy - h / 2) / this.scale + this.cy]
  }

  panPx(dx: number, dy: number) {
    this.cx -= dx / this.scale
    this.cy -= dy / this.scale
  }

  zoomAt(sx: number, sy: number, w: number, h: number, factor: number) {
    const [wx, wy] = this.toWorld(sx, sy, w, h)
    this.scale = Math.min(20, Math.max(0.02, this.scale * factor))
    // Keep the world point under the cursor fixed on screen.
    this.cx = wx - (sx - w / 2) / this.scale
    this.cy = wy - (sy - h / 2) / this.scale
  }

  /** Center + zoom so the whole map is visible with a margin. */
  fit(geometry: MapGeometry, w: number, h: number) {
    const xs: number[] = []
    const ys: number[] = []
    for (const n of geometry.nodes) { xs.push(n.x); ys.push(n.y) }
    for (const b of geometry.buildings) { xs.push(b.x); ys.push(b.y) }
    if (xs.length === 0) return
    const minX = Math.min(...xs), maxX = Math.max(...xs)
    const minY = Math.min(...ys), maxY = Math.max(...ys)
    this.cx = (minX + maxX) / 2
    this.cy = (minY + maxY) / 2
    const spanX = Math.max(maxX - minX, 100)
    const spanY = Math.max(maxY - minY, 100)
    this.scale = Math.min(20, Math.max(0.02,
      0.85 * Math.min(w / spanX, h / spanY)))
  }
}
